### PR TITLE
New last/lastdb module to index sequences before alignment.

### PR DESCRIPTION
**Pull request by charles-plessy**
The `lastdb` command creates a sequence index for the [LAST aligner](https://gitlab.com/mcfrith/last). Input can be in FASTA or FASTQ format, and compression is handled automagically.  DNA or protein sequences can be indexed.

The sequence index is a collection of files sharing the same basename. This module sets the basename to the sample identifier (`$meta.id`) and creates the index in a directory always called `lastdb`.  The module's output channel then conveys a copy of the metadata and the path to the `lastdb` directory.

Other modules will follow (see Issue #464).  The LAST aligner can align proteins to proteins, DNA to DNA and can translate DNA align to proteins.

## PR checklist

Closes #XXX 

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `<SOFTWARE>.version.txt` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

File: tests/config/pytest_software.yml [hunk 1]
```diff
@@ -334,6 +334,10 @@ kraken2/run:
   - software/untar/**
   - tests/software/kraken2/run/**
 
+last_lastdb:
```
Comment (drpatelh) [inline tests/config/pytest_software.yml:337]:
```suggestion
last/lastdb:
```

File: tests/software/last/lastdb/test.yml [hunk 2]
```diff
@@ -0,0 +1,42 @@
+- name: last lastdb test_last_lastdb
+  command: nextflow run tests/software/last/lastdb -entry test_last_lastdb -c tests/config/nextflow.config
+  tags:
+    - last
+    - last_lastdb
```
Comment (drpatelh) [inline tests/software/last/lastdb/test.yml:5]:
```suggestion
    - last/lastdb
```

File: tests/software/last/lastdb/test.yml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      md5sum: 4137fb6fe9df2b3d78d5b960390aac7b
+    - path: output/last/lastdb/test.suf
+      md5sum: 1895efa8653e8e9bd3605cff0408ed33
+    - path: output/last/lastdb/test.tis
+      md5sum: b7c40f06b1309dc6f37849eeb86dfd22
+
+- name: last lastdb test_last_lastdb_gzipped_input
+  command: nextflow run tests/software/last/lastdb -entry test_last_lastdb_gzipped_input -c tests/config/nextflow.config
+  tags:
+    - last_lastdb
```
Comment (drpatelh) [inline tests/software/last/lastdb/test.yml:25]:
```suggestion
    - last/lastdb
```

File: software/last/lastdb/main.nf [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    input:
+    tuple val(meta), path(fastx)
+
+    output:
+    tuple val(meta), path("lastdb"), emit: index
+    path "*.version.txt"           , emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+
```
Comment (drpatelh) [inline software/last/lastdb/main.nf:30]:
```suggestion
    def software = getSoftwareName(task.process)
    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
```

File: software/last/lastdb/main.nf [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    """
+    mkdir lastdb
+    lastdb \\
+        ${options.args} \\
+        -P ${task.cpus} \\
+        lastdb/${meta.id} \\
```
Comment (drpatelh) [inline software/last/lastdb/main.nf:36]:
```suggestion
        lastdb/${prefix} \\
```

File: software/last/lastdb/main.nf [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    tuple val(meta), path("lastdb"), emit: index
+    path "*.version.txt"           , emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+
+    """
+    mkdir lastdb
+    lastdb \\
+        ${options.args} \\
```
Comment (drpatelh) [inline software/last/lastdb/main.nf:34]:
```suggestion
        $options.args \\
```

File: software/last/lastdb/main.nf [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        -P ${task.cpus} \\
```
Comment (drpatelh) [inline software/last/lastdb/main.nf:35]:
```suggestion
        -P $task.cpus \\
```

File: software/last/lastdb/main.nf [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        lastdb/${meta.id} \\
+        ${fastx}
```
Comment (drpatelh) [inline software/last/lastdb/main.nf:37]:
```suggestion
        $fastx
```

File: software/last/lastdb/main.nf [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+params.options = [:]
+options        = initOptions(params.options)
+
+process LAST_LASTDB {
+    tag "$meta.id"
+    label 'process_medium'
+    publishDir "${params.outdir}",
+        mode: params.publish_dir_mode,
+        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), publish_id:meta.id) }
```
Comment (drpatelh) [inline software/last/lastdb/main.nf:12]:
```suggestion
        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), meta:meta, publish_by_meta:['id']) }
```

File: software/last/lastdb/functions.nf [hunk 10]
```diff
@@ -0,0 +1,60 @@
+/*
+ * -----------------------------------------------------
```
Comment (drpatelh) [inline software/last/lastdb/functions.nf:2]: This file needs to be replace with [this one](https://raw.githubusercontent.com/nf-core/modules/master/software/lib/functions.nf). It was updated recently.

**Review by drpatelh — CHANGES_REQUESTED** [review]
Looks awesome overall @charles-plessy 😍 Mainly minor changes that are possibly because you need to use the `dev` version of nf-core/tools to create the module file at the moment. You can do this by installing it with the command:

```console
pip install --upgrade --force-reinstall git+https://github.com/nf-core/tools.git@dev
nf-core modules create
```

**Review by charles-plessy — COMMENTED** [review]
Thanks for the changes: I accepted all and updated `functions.nf`

File: tests/software/last/lastdb/test.yml [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+      md5sum: 1895efa8653e8e9bd3605cff0408ed33
+    - path: output/last/lastdb/test.tis
+      md5sum: b7c40f06b1309dc6f37849eeb86dfd22
+
+- name: last lastdb test_last_lastdb_gzipped_input
+  command: nextflow run tests/software/last/lastdb -entry test_last_lastdb_gzipped_input -c tests/config/nextflow.config
+  tags:
+    - last/lastdb
+    - last_lastdb_gzipped_input
+    - last
```
Comment (drpatelh) [inline tests/software/last/lastdb/test.yml:27]:
```suggestion
    - last
    - last/lastdb
```

Comment (drpatelh) [reply tests/software/last/lastdb/test.yml:27]: Sorry, can't push commits to your fork @charles-plessy but would you mind making this change too? I think that is why the linting is failing.

**Review by drpatelh — APPROVED** [review]

**Comment by drpatelh** [comment]
💥 Thanks @charles-plessy !

**Pull request merged**